QEP-Fields-and-forms-redesign.rst

```diff
@@ -0,0 +1,133 @@
+.. _qep#[.#]:
+
+========================================================================
+QGIS Enhancement #: Fields And Forms Redesign
+========================================================================
+
+:Date: 2014/11/25
+:Author: Matthias Kuhn
+:Contact: matthias dot kuhn at gmx dot ch
+:Last Edited: 2014/11/25
+:Status:  Draft
+:Version: QGIS 2.8
+
+1. Summary
+----------
+
+The tab *Fields* on the vector layer properties dialog should be split into
+two different tabs that separate data-related from form- and widget-related
+user interface elements.
+This is expected to improve user experience and usability.
+
+2. Rationale
+----------
+
+On the vector layer properties dialog there is a tab *Fields* where the user is
+able to configure different things related to fields (data- and widget-related)
+and forms (data- and form-related):
+
+- The general layout of the form (autogenerated, drag and drop, ui file)
+- The appearance of the individual widgets
+
+  + Some weak constraints can be introduced that will limit acceptable input
+    from widgets but will not be enforced by any other way to set a field's
+    value (e.g. min/max values from the range widget)
+    
+- Adding fields and removing fields
+- Drag and drop fields and relations to the form
+- Python logic
+
+This tab currently is in a very bad shape from a UX perspective. Problems
+include:
+
+- Having to scroll horizontally in the fields table when wanting to change
+  the widget type or looking for information.
+- The resizable containers on the right and at the bottom can be hidden
+  (consciously and unconsciously) and confuse the user.
+
+3. Proposal
+-------------
+
+The information and functionality that is now available on the _Fields_ tab
+will be split and put on two different tabs:
+
+- Fields
+- Form
+
+3.1 Fields
+.................
+
+The *Fields* tab conatins anything that is related to data and fields.
+
+In particular, this refers to:
+
+- Changing fields
+
+  + add
+  + delete
+  + calculate
+
+- Information about fields
+
+  + data type
+  + comment
+  + constraints (Future)
+  + domain (Future)
+  + exposure (WMS/WFS)
+  + ...
+
+Mockup
+~~~~~~~~~~~~~
+
+.. figure:: fields.png
+   
+   Mockup of the fields tab. This will be shown embedded in the vector layer properties dialog.
+
+3.2 Forms
+.................
+
+The *Form* tab contains anything that is related to user interface and visual
+representation and manipulation of a features fields.
+
+In particular, this refers to:
+
+- The widget type
+- The widget configuration
+- Form layout
+- Python init script for the form
+- ...
+
+Mockups
+~~~~~~~~~~~~~
+
+.. figure:: form1.png
+
+   Mockup of the *Form* tab with autogenerated layout. This will be shown embedded in the vector layer properties dialog.
```

Comment:
Sorry, forget about this, there's no need to specify NULL as a default value, it's already the result when nothing is chosen.  It's stupid.  I was confused because with dropdown lists, even when 'Allow null value' is not checked, it is still possible to chose nothing (by simply not touching the dropdown) and end up with a NULL ! This is not coherent and I will file a ticket for it. 

But it would be very useful the set a field as mandatory, meaning that something has to be entered, and therefore there should be a control when saving the form.